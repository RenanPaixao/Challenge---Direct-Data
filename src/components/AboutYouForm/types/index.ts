export interface FieldConfig<t> {
  name: keyof t,
  type: string,
  label: string
}
