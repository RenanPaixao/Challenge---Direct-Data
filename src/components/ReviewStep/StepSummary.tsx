import { Box, BoxProps, Flex, Heading, Text } from '@chakra-ui/react'

interface SummarySection {
  sectionTitle: string
  sectionContent: {
    [key: string]: string
  }
}
interface IProps extends BoxProps, SummarySection{
  nested?: SummarySection[]
}

export const StepSummary = ({ sectionTitle, sectionContent, nested, ...rest }: IProps) => {
  const entries = Object.entries(sectionContent)

  return <Box {...rest}>
    <Heading as={'h3'} size={'md'} textAlign={'start'} >{sectionTitle}</Heading>
    <Box p={8} mt={4} border={'thin solid var(--chakra-colors-gray-300)'} borderRadius={8}>
      {
        entries.map(([key, value]) => (
          <Flex key={key}>
            <Text aria-label={`${sectionTitle} - ${key}`}><strong>{key}</strong></Text>: {value}
          </Flex>
        ))
      }
      {
        nested && nested.map(({ sectionTitle, sectionContent }) => (
          <>
            <Heading as={'h4'} color={'var(--chakra-colors-purple-500)'} mt={8} mb={2} size={'sm'} textAlign={'start'}>
              <strong>{sectionTitle}</strong>
            </Heading>
            {
              Object.entries(sectionContent).map(([key, value]) => (
                <Flex key={key}>
                  <Text aria-label={`${sectionTitle} - ${key}`}><strong>{key}</strong></Text>: {value}
                </Flex>
              ))
            }
          </>
        ))
      }
    </Box>
  </Box>
}
