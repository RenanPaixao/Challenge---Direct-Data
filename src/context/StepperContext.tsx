import { UseStepsReturn } from '@chakra-ui/react'
import { createContext, PropsWithChildren } from 'react'
import { IStep } from '../components/TheStepper/TheStepper.tsx'

interface IProps extends UseStepsReturn {
  steps: IStep[]
}
export const StepperContext = createContext<IProps | null>(null)

StepperContext.displayName = 'StepperContext'

export const StepperProvider = ({ children, ...rest }: PropsWithChildren<IProps>) => {
  return <StepperContext.Provider value={{ ...rest }}>
    {children}
  </StepperContext.Provider>
}

