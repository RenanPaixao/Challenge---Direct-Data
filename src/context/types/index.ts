
export interface AboutYouInformation extends IdentityInformation {
  responsible: IdentityInformation | null
  weight: string
  height: string
  email: string
  phone: string
}

export interface Address {
  street: string,
  number: string,
  complement?: string,
  district: string,
  cep: string,
  city: string,
  state: string
}

export interface IdentityInformation {
  name: string ,
  lastName: string ,
  birthDate: string,
  cpf: string
}
