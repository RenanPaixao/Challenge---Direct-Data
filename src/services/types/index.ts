export interface StudentModel extends IdentityInformationModel {
  responsavel: IdentityInformationModel | null,
  peso: number,
  altura: number,
  email: string,
  telefoneContato: string,
  logradouro: string,
  numero: number,
  complemento: string,
  bairro: string,
  cep: string,
  cidade: string,
  uf: string
}

interface IdentityInformationModel {
  nome: string ,
  sobrenome: string ,
  dataNascimento: string,
  cpf: string
}
