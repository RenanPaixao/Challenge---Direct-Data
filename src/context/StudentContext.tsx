import { createContext, PropsWithChildren, useEffect, useState } from 'react'
import { AboutYouInformation } from './types'
import { persistOnSessionStorage, retrieveFromSessionStorage } from '../utils/persistence.ts'

export interface IProps {
  aboutYouInformation: AboutYouInformation | null
  setAboutYouInformation: (aboutYouInformation: AboutYouInformation) => void
}

export const StudentContext = createContext<IProps>({
  aboutYouInformation: null,
  setAboutYouInformation: () => {}
})

StudentContext.displayName = 'StudentContext'

export const StudentProvider = ({ children }: PropsWithChildren) => {
  const sessionStorageKey = 'aboutYouInformation'
  const [aboutYouInformation, setAboutYouInformation] = useState<AboutYouInformation | null>(() => {
    return retrieveFromSessionStorage(sessionStorageKey)
  })

  useEffect(() => {
    persistOnSessionStorage(sessionStorageKey, aboutYouInformation)
  }, [aboutYouInformation])

  return <StudentContext.Provider value={{
    aboutYouInformation,
    setAboutYouInformation
  }}>
    {children}
  </StudentContext.Provider>
}
