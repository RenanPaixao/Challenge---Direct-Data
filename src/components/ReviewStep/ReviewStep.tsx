import { useContext } from 'react'
import { StudentContext } from '../../context/StudentContext.tsx'
import { Box, Button, Center } from '@chakra-ui/react'
import { StepSummary } from './StepSummary.tsx'
import { FormFooterWrapper } from '../FormFooterWrapper/FormFooterWrapper.tsx'
import { StepperContext } from '../../context/StepperContext.tsx'
import { DateTime } from 'luxon'
import { useNavigate } from 'react-router'

export const ReviewStep = () => {
  const { goToPrevious } = useContext(StepperContext)!
  const { address, aboutYouInformation } = useContext(StudentContext)
  const navigate = useNavigate()
  const { responsible, ...aboutYouWithoutResponsible } = aboutYouInformation!

  const aboutYouWithFormattedDate = formatBirthDate(aboutYouWithoutResponsible)
  const responsibleWithFormattedDate = formatBirthDate(responsible ?? {})

  const sectionSteps = [
    {
      sectionTitle: 'Sobre Você',
      sectionContent: mapObjectKeysToPortuguese(aboutYouWithFormattedDate),
      nested: [
        {
          sectionTitle: 'Responsável',
          sectionContent: mapObjectKeysToPortuguese(responsibleWithFormattedDate)
        }
      ]
    },
    {
      sectionTitle: 'Endereço',
      sectionContent: mapObjectKeysToPortuguese(address ?? {})
    }
  ]

  /**
   * Navigate to the success page.
   */
  function goToSuccessPage() {
    navigate('/success')
  }

  return <Center flexDirection={'column'}>
    <Box w={'75%'} minW={'600px'}>
      {
        sectionSteps.map(({ sectionContent, sectionTitle, nested }) => (
          <StepSummary mt={4} key={sectionTitle} sectionTitle={sectionTitle} sectionContent={sectionContent} nested={nested}/>
        ))}
    </Box>
    <FormFooterWrapper mt={8}>
      <Button onClick={goToPrevious}>Voltar</Button>
      <Button onClick={goToSuccessPage}>Avançar</Button>
    </FormFooterWrapper>
  </Center>
}

/**
 * Map object keys to Portuguese.
 * @param object
 */
function mapObjectKeysToPortuguese(object: Record<string, any>) {
  const allKeys: Record<string, string> = {
    name: 'Nome',
    lastName: 'Sobrenome',
    birthDate: 'Data de nascimento',
    cpf: 'CPF',
    weight: 'Peso',
    height: 'Altura',
    email: 'Email',
    phone: 'Telefone',
    responsibleName: 'Nome do responsável',
    responsibleLastName: 'Sobrenome do responsável',
    responsibleBirthDate: 'Data de nascimento do responsável',
    responsibleCpf: 'CPF do responsável',
    responsibleEmail: 'Email do responsável',
    responsiblePhone: 'Telefone do responsável',
    cep: 'CEP',
    street: 'Rua',
    number: 'Número',
    complement: 'Complemento',
    district: 'Bairro',
    city: 'Cidade',
    state: 'Estado'
  }
  return Object.keys(object).reduce((acc, key) => {
    acc[allKeys[key]] = object[key]
    return acc
  }, {} as Record<string, string>)
}

/**
 * Format the birthdate of an object.
 */
function formatBirthDate(object: Record<string, any>) {
  return {
    ...object,
    birthDate: DateTime.fromISO(object.birthDate).toFormat('dd/MM/yyyy')
  }
}
