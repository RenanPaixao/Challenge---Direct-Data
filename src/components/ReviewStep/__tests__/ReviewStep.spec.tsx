
import { renderWithStudentContext } from '../../../../tests/test-utils.tsx'
import { ReviewStep } from '../ReviewStep.tsx'
import { AboutYouInformation, Address } from '../../../context/types'
import { expect } from 'vitest'

const aboutYouValues: AboutYouInformation = {
  name: 'test name',
  lastName: 'test last name',
  birthDate: '2015-01-01',
  cpf: '123.456.789-00',
  email: 'test@test.com',
  height: '1.80',
  weight: '70',
  phone: '1234567890',
  responsible: {
    name: 'test responsible name',
    cpf: '123.456.789-00',
    birthDate: '2000-02-03',
    lastName: 'test responsible last name'
  }
}

const addressValues: Address = {
  cep: '12345-123',
  number: '12',
  city: 'test city',
  state: 'al',
  complement: 'test complement',
  street: 'test street',
  district: 'test district'
}

describe('ReviewStep', () => {
  it('should render', async () => {
    const { container } = renderWithStudentContext(<ReviewStep/>, {
      address: addressValues,
      aboutYouInformation: aboutYouValues
    })

    expect(container).toMatchSnapshot()
  })

  it('should render with formatted birthDate', async () => {
    const { getByText } = renderWithStudentContext(<ReviewStep/>, {
      address: addressValues,
      aboutYouInformation: aboutYouValues
    })

    const studentBirthDate = getByText(': 01/01/2015')
    const responsibleBirthDate = getByText(': 03/02/2000')

    expect(studentBirthDate).toBeTruthy()
    expect(responsibleBirthDate).toBeTruthy()
  })
})
