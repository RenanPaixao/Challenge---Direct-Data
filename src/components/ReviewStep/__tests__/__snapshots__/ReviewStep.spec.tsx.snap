// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ReviewStep > should render 1`] = `
<div>
  <div
    class="css-b93e9j"
  >
    <div
      class="css-1lu5lqd"
    >
      <div
        class="css-1ayfwcb"
      >
        <h3
          class="chakra-heading css-2ghjnd"
        >
          Sobre Você
        </h3>
        <div
          class="css-1lcfq1r"
        >
          <div
            class="css-k008qs"
          >
            <p
              aria-label="Sobre Você - Nome"
              class="chakra-text css-0"
            >
              <strong>
                Nome
              </strong>
            </p>
            : 
            test name
          </div>
          <div
            class="css-k008qs"
          >
            <p
              aria-label="Sobre Você - Sobrenome"
              class="chakra-text css-0"
            >
              <strong>
                Sobrenome
              </strong>
            </p>
            : 
            test last name
          </div>
          <div
            class="css-k008qs"
          >
            <p
              aria-label="Sobre Você - CPF"
              class="chakra-text css-0"
            >
              <strong>
                CPF
              </strong>
            </p>
            : 
            123.456.789-00
          </div>
          <div
            class="css-k008qs"
          >
            <p
              aria-label="Sobre Você - Email"
              class="chakra-text css-0"
            >
              <strong>
                Email
              </strong>
            </p>
            : 
            test@test.com
          </div>
          <div
            class="css-k008qs"
          >
            <p
              aria-label="Sobre Você - Altura"
              class="chakra-text css-0"
            >
              <strong>
                Altura
              </strong>
            </p>
            : 
            1.80
          </div>
          <div
            class="css-k008qs"
          >
            <p
              aria-label="Sobre Você - Peso"
              class="chakra-text css-0"
            >
              <strong>
                Peso
              </strong>
            </p>
            : 
            70
          </div>
          <div
            class="css-k008qs"
          >
            <p
              aria-label="Sobre Você - Data de nascimento"
              class="chakra-text css-0"
            >
              <strong>
                Data de nascimento
              </strong>
            </p>
            : 
            01/01/2015
          </div>
          <div
            class="css-k008qs"
          >
            <p
              aria-label="Sobre Você - Telefone"
              class="chakra-text css-0"
            >
              <strong>
                Telefone
              </strong>
            </p>
            : 
            1234567890
          </div>
          <h4
            class="chakra-heading css-mhi7l6"
          >
            <strong>
              Responsável
            </strong>
          </h4>
          <div
            class="css-k008qs"
          >
            <p
              aria-label="Responsável - Nome"
              class="chakra-text css-0"
            >
              <strong>
                Nome
              </strong>
            </p>
            : 
            test responsible name
          </div>
          <div
            class="css-k008qs"
          >
            <p
              aria-label="Responsável - CPF"
              class="chakra-text css-0"
            >
              <strong>
                CPF
              </strong>
            </p>
            : 
            123.456.789-00
          </div>
          <div
            class="css-k008qs"
          >
            <p
              aria-label="Responsável - Data de nascimento"
              class="chakra-text css-0"
            >
              <strong>
                Data de nascimento
              </strong>
            </p>
            : 
            03/02/2000
          </div>
          <div
            class="css-k008qs"
          >
            <p
              aria-label="Responsável - Sobrenome"
              class="chakra-text css-0"
            >
              <strong>
                Sobrenome
              </strong>
            </p>
            : 
            test responsible last name
          </div>
        </div>
      </div>
      <div
        class="css-1ayfwcb"
      >
        <h3
          class="chakra-heading css-2ghjnd"
        >
          Endereço
        </h3>
        <div
          class="css-1lcfq1r"
        >
          <div
            class="css-k008qs"
          >
            <p
              aria-label="Endereço - CEP"
              class="chakra-text css-0"
            >
              <strong>
                CEP
              </strong>
            </p>
            : 
            12345-123
          </div>
          <div
            class="css-k008qs"
          >
            <p
              aria-label="Endereço - Bairro"
              class="chakra-text css-0"
            >
              <strong>
                Bairro
              </strong>
            </p>
            : 
            test district
          </div>
          <div
            class="css-k008qs"
          >
            <p
              aria-label="Endereço - Rua"
              class="chakra-text css-0"
            >
              <strong>
                Rua
              </strong>
            </p>
            : 
            test street
          </div>
          <div
            class="css-k008qs"
          >
            <p
              aria-label="Endereço - Complemento"
              class="chakra-text css-0"
            >
              <strong>
                Complemento
              </strong>
            </p>
            : 
            test complement
          </div>
          <div
            class="css-k008qs"
          >
            <p
              aria-label="Endereço - Estado"
              class="chakra-text css-0"
            >
              <strong>
                Estado
              </strong>
            </p>
            : 
            al
          </div>
          <div
            class="css-k008qs"
          >
            <p
              aria-label="Endereço - Cidade"
              class="chakra-text css-0"
            >
              <strong>
                Cidade
              </strong>
            </p>
            : 
            test city
          </div>
          <div
            class="css-k008qs"
          >
            <p
              aria-label="Endereço - Número"
              class="chakra-text css-0"
            >
              <strong>
                Número
              </strong>
            </p>
            : 
            12
          </div>
        </div>
      </div>
    </div>
    <div
      class="css-2srqce"
    >
      <button
        class="chakra-button css-ez23ye"
        type="button"
      >
        Voltar
      </button>
      <button
        class="chakra-button css-ez23ye"
        type="button"
      >
        Avançar
      </button>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
